// rumble-charts/lib/helpers/transforms/sort.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = sort;

var _defaults2 = _interopRequireDefault(require("lodash/defaults"));

var _sortBy2 = _interopRequireDefault(require("lodash/sortBy"));

var _map2 = _interopRequireDefault(require("lodash/map"));

var _isString2 = _interopRequireDefault(require("lodash/isString"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function sort(props, options) {
  var _ref = options || {},
      direction = _ref.direction;

  direction = (0, _isString2["default"])(direction) ? direction.trim().toLowerCase() : 'asc';
  var series = props.series,
      seriesNormalized = props.seriesNormalized,
      maxX = props.maxX,
      maxY = props.maxY,
      minX = props.minX,
      minY = props.minY;
  series = (0, _map2["default"])(series, function (series) {
    var newSeries = {};
    newSeries.data = (0, _sortBy2["default"])(series.data, 'y');

    if (direction === 'desc') {
      newSeries.data.reverse();
    }

    newSeries.data = (0, _map2["default"])(newSeries.data, function (point, pointIndex) {
      var newPoint = {
        realX: point.x,
        x: pointIndex
      };
      return (0, _defaults2["default"])(newPoint, point);
    });
    newSeries.data = (0, _sortBy2["default"])(newSeries.data, 'realX');
    return (0, _defaults2["default"])(newSeries, series);
  });
  return {
    series: series,
    seriesNormalized: seriesNormalized,
    maxX: maxX,
    maxY: maxY,
    minX: minX,
    minY: minY
  };
}
