// rumble-charts/lib/helpers/transforms/transpose.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = transpose;

var _defaults2 = _interopRequireDefault(require("lodash/defaults"));

var _forEach2 = _interopRequireDefault(require("lodash/forEach"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function transpose(props) {
  var series = props.series,
      seriesNormalized = props.seriesNormalized,
      minX = props.minX,
      maxY = props.maxY,
      minY = props.minY;
  var maxX = 0;
  var newSeries = [];
  (0, _forEach2["default"])(series, function (series, seriesIndex) {
    (0, _forEach2["default"])(series.data, function (point, pointIndex) {
      newSeries[pointIndex] = newSeries[pointIndex] || {
        data: []
      };
      maxX = Math.max(maxX, seriesIndex);
      newSeries[pointIndex].data[seriesIndex] = (0, _defaults2["default"])({
        realX: point.x,
        x: seriesIndex
      }, point);
    });
  });
  return {
    series: newSeries,
    seriesNormalized: seriesNormalized,
    maxX: maxX,
    maxY: maxY,
    minX: minX,
    minY: minY
  };
}
