// rumble-charts/lib/helpers/transforms/unstack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = unstack;

var _defaults2 = _interopRequireDefault(require("lodash/defaults"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _map2 = _interopRequireDefault(require("lodash/map"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function unstack(props) {
  var series = props.series,
      seriesNormalized = props.seriesNormalized,
      maxX = props.maxX,
      maxY = props.maxY,
      minX = props.minX,
      minY = props.minY;
  var newSeries = (0, _map2["default"])(series, function (series) {
    var newSeries = {
      data: (0, _map2["default"])(series.data, function (point) {
        var newPoint = (0, _omit2["default"])(point, ['y0', 'y']);
        newPoint.y = point.y - point.y0;
        return newPoint;
      })
    };
    return (0, _defaults2["default"])(newSeries, series);
  }, []);
  return {
    series: newSeries,
    seriesNormalized: seriesNormalized,
    maxX: maxX,
    maxY: maxY,
    minX: minX,
    minY: minY
  };
}
