// rumble-charts/lib/DropShadow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = DropShadow;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function DropShadow(props) {
  return _react["default"].createElement("filter", {
    id: props.id
  }, _react["default"].createElement("feGaussianBlur", {
    "in": props.blurIn,
    stdDeviation: props.blurDeviation
  }), _react["default"].createElement("feOffset", {
    dx: props.dx,
    dy: props.dy
  }), _react["default"].createElement("feMerge", null, _react["default"].createElement("feMergeNode", null), _react["default"].createElement("feMergeNode", {
    "in": 'SourceGraphic'
  })));
}

DropShadow.displayName = 'DropShadow';
DropShadow.propTypes = {
  id: _propTypes["default"].string.isRequired,
  dx: _propTypes["default"].number,
  dy: _propTypes["default"].number,
  blurDeviation: _propTypes["default"].number,
  blurIn: _propTypes["default"].oneOf(['SourceAlpha', 'SourceGraphic', 'BackgroundImage', 'BackgroundAlpha', 'FillPaint', 'StrokePaint'])
};
DropShadow.defaultProps = {
  dx: 1,
  dy: 1,
  blurDeviation: 4,
  blurIn: 'SourceAlpha'
};
