// rumble-charts/lib/Labels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Labels;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _propTypes2 = _interopRequireDefault(require("./helpers/propTypes"));

var _Dots = _interopRequireDefault(require("./Dots"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/**
 * Renders labels for dots. Internally it's just a wrapper for [`<Dots>`](#dots) component
 * with `dotType="circle"`.
 *
 * @example ../docs/examples/Labels.md
 */
function Labels(props) {
  return _react["default"].createElement(_Dots["default"], _extends({}, props, {
    dotType: 'labels'
  }));
}

Labels.displayName = 'Labels';
Labels.propTypes = {
  className: _propTypes["default"].string,
  colors: _propTypes["default"].oneOfType([_propTypes["default"].oneOf(['category10', 'category20', 'category20b', 'category20c', 'accent', 'dark2', 'paired', 'pastel1', 'pastel2', 'set1', 'set2', 'set3', 'tableau10']), _propTypes["default"].arrayOf(_propTypes["default"].string), _propTypes["default"].func]),
  opacity: _propTypes["default"].number,
  style: _propTypes["default"].object,
  label: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].func]),
  labelFormat: _propTypes["default"].func,
  labelAttributes: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func]),
  seriesVisible: _propTypes["default"].oneOfType([_propTypes["default"].bool, _propTypes["default"].func]),
  seriesAttributes: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func]),
  seriesStyle: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func]),
  groupStyle: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func]),
  dotVisible: _propTypes["default"].oneOfType([_propTypes["default"].bool, _propTypes["default"].func]),
  dotAttributes: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func]),
  dotStyle: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func]),
  seriesIndex: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].array, _propTypes["default"].func]),
  series: _propTypes2["default"].series
};
Labels.defaultProps = {
  colors: 'category20',
  circleRadius: 4,
  ellipseRadiusX: 6,
  ellipseRadiusY: 4,
  seriesVisible: true,
  dotVisible: true
};
