// rumble-charts/lib/Title.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Title;

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _isString2 = _interopRequireDefault(require("lodash/isString"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _getCoords2 = _interopRequireDefault(require("./helpers/getCoords"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * Renders title (text or everything else) at specified `position`.
 *
 * @example ../docs/examples/Title.md
 */
function Title(props) {
  var style = props.style,
      className = props.className,
      children = props.children;

  var _getCoords = (0, _getCoords2["default"])(props.position, props.layerWidth, props.layerHeight, props.width, props.height),
      x = _getCoords.x,
      y = _getCoords.y;

  return _react["default"].createElement("g", {
    className: className,
    transform: 'translate(' + x + ' ' + y + ')',
    style: style
  }, (0, _isString2["default"])(children) ? _react["default"].createElement("text", null, children) : (0, _isFunction2["default"])(children) ? children(props) : children);
}

Title.displayName = 'Title';
Title.propTypes = {
  className: _propTypes["default"].string,
  style: _propTypes["default"].object,
  width: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  height: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  position: _propTypes["default"].oneOfType([_propTypes["default"].array, _propTypes["default"].string]),
  children: _propTypes["default"].oneOfType([_propTypes["default"].func, _propTypes["default"].string, _propTypes["default"].node]).isRequired,
  layerWidth: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  layerHeight: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  series: _propTypes["default"].array
};
Title.defaultProps = {
  position: 'top center'
};
