// rumble-charts/lib/Transform.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Transform;

var _isUndefined2 = _interopRequireDefault(require("lodash/isUndefined"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _transform = _interopRequireDefault(require("./helpers/transform"));

var _proxyChildren = _interopRequireDefault(require("./helpers/proxyChildren"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * Transforms `series` data according chosen `method`.
 *
 * As a wrapper it takes `series` obtained from its parent and gives it to its children.
 *
 * @example ../docs/examples/Transform.md
 */
function Transform(props) {
  var className = props.className,
      layerWidth = props.layerWidth,
      layerHeight = props.layerHeight,
      scaleX = props.scaleX,
      scaleY = props.scaleY;
  var newProps = (0, _transform["default"])(props, props.method);
  var children = (0, _proxyChildren["default"])(props.children, newProps, {
    layerWidth: (0, _isUndefined2["default"])(newProps.layerWidth) ? layerWidth : newProps.layerWidth,
    layerHeight: (0, _isUndefined2["default"])(newProps.layerHeight) ? layerHeight : newProps.layerHeight,
    scaleX: (0, _isUndefined2["default"])(newProps.scaleX) ? scaleX : newProps.scaleX,
    scaleY: (0, _isUndefined2["default"])(newProps.scaleY) ? scaleY : newProps.scaleY
  });
  return _react["default"].createElement("g", {
    className: className
  }, children);
}

Transform.displayName = 'Transform';
Transform.propTypes = {
  className: _propTypes["default"].string,
  series: _propTypes["default"].array,

  /**
   * Possible string values: "`stack`", "`stackNormalized`", "`sort`", "`unstack`",
   * "`transpose`", "`rotate`", "`reverse`". Also you can define it as function that
   * receives props as an object, transforms it somehow and returns changes props as an object.
   */
  method: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].func, _propTypes["default"].array]),
  scaleX: _propTypes["default"].object,
  scaleY: _propTypes["default"].object,
  layerWidth: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  layerHeight: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  children: _propTypes["default"].node
};
