// rumble-charts/lib/helpers/colorFunc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = colorFunc;
exports.defaultSchemeName = exports.defaultScheme = exports.schemeMap = void 0;

var _isString2 = _interopRequireDefault(require("lodash/isString"));

var _isEmpty2 = _interopRequireDefault(require("lodash/isEmpty"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _d3Scale = require("d3-scale");

var _d3ScaleChromatic = require("d3-scale-chromatic");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var schemeMap = {
  category10: _d3ScaleChromatic.schemeCategory10,
  category20: ['#1f77b4', '#aec7e8', '#ff7f0e', '#ffbb78', '#2ca02c', '#98df8a', '#d62728', '#ff9896', '#9467bd', '#c5b0d5', '#8c564b', '#c49c94', '#e377c2', '#f7b6d2', '#7f7f7f', '#c7c7c7', '#bcbd22', '#dbdb8d', '#17becf', '#9edae5'],
  category20b: ['#393b79', '#5254a3', '#6b6ecf', '#9c9ede', '#637939', '#8ca252', '#b5cf6b', '#cedb9c', '#8c6d31', '#bd9e39', '#e7ba52', '#e7cb94', '#843c39', '#ad494a', '#d6616b', '#e7969c', '#7b4173', '#a55194', '#ce6dbd', '#de9ed6'],
  category20c: ['#3182bd', '#6baed6', '#9ecae1', '#c6dbef', '#e6550d', '#fd8d3c', '#fdae6b', '#fdd0a2', '#31a354', '#74c476', '#a1d99b', '#c7e9c0', '#756bb1', '#9e9ac8', '#bcbddc', '#dadaeb', '#636363', '#969696', '#bdbdbd', '#d9d9d9'],
  accent: _d3ScaleChromatic.schemeAccent,
  dark2: _d3ScaleChromatic.schemeDark2,
  paired: _d3ScaleChromatic.schemePaired,
  pastel1: _d3ScaleChromatic.schemePastel1,
  pastel2: _d3ScaleChromatic.schemePastel2,
  set1: _d3ScaleChromatic.schemeSet1,
  set2: _d3ScaleChromatic.schemeSet2,
  set3: _d3ScaleChromatic.schemeSet3,
  tableau10: _d3ScaleChromatic.schemeTableau10
};
exports.schemeMap = schemeMap;
var defaultScheme = schemeMap.category20;
exports.defaultScheme = defaultScheme;
var defaultSchemeName = 'category20';
exports.defaultSchemeName = defaultSchemeName;

function colorFunc(colors) {
  if ((0, _isFunction2["default"])(colors)) {
    return colors;
  } else if ((0, _isEmpty2["default"])(colors)) {
    return (0, _d3Scale.scaleOrdinal)(defaultScheme);
  } else if ((0, _isString2["default"])(colors)) {
    return (0, _d3Scale.scaleOrdinal)(schemeMap[colors]);
  } else {
    return (0, _d3Scale.scaleOrdinal)(colors);
  }
}
