// rumble-charts/lib/helpers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.transforms = exports.value = exports.transform = exports.proxyChildren = exports.propTypes = exports.normalizeSeries = exports.normalizeNumber = exports.getCoords = exports.eases = exports.curves = exports.colorFunc = void 0;

var _colorFunc2 = _interopRequireDefault(require("./colorFunc"));

var _curves2 = _interopRequireDefault(require("./curves"));

var _eases2 = _interopRequireDefault(require("./eases"));

var _getCoords2 = _interopRequireDefault(require("./getCoords"));

var _normalizeNumber2 = _interopRequireDefault(require("./normalizeNumber"));

var _normalizeSeries2 = _interopRequireDefault(require("./normalizeSeries"));

var _propTypes2 = _interopRequireDefault(require("./propTypes"));

var _proxyChildren2 = _interopRequireDefault(require("./proxyChildren"));

var _transform2 = _interopRequireDefault(require("./transform"));

var _value2 = _interopRequireDefault(require("./value"));

var _transforms2 = _interopRequireDefault(require("./transforms"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var colorFunc = _colorFunc2["default"];
exports.colorFunc = colorFunc;
var curves = _curves2["default"];
exports.curves = curves;
var eases = _eases2["default"];
exports.eases = eases;
var getCoords = _getCoords2["default"];
exports.getCoords = getCoords;
var normalizeNumber = _normalizeNumber2["default"];
exports.normalizeNumber = normalizeNumber;
var normalizeSeries = _normalizeSeries2["default"];
exports.normalizeSeries = normalizeSeries;
var propTypes = _propTypes2["default"];
exports.propTypes = propTypes;
var proxyChildren = _proxyChildren2["default"];
exports.proxyChildren = proxyChildren;
var transform = _transform2["default"];
exports.transform = transform;
var value = _value2["default"];
exports.value = value;
var transforms = _transforms2["default"];
exports.transforms = transforms;
var _default = {
  colorFunc: colorFunc,
  curves: curves,
  eases: eases,
  getCoords: getCoords,
  normalizeNumber: normalizeNumber,
  normalizeSeries: normalizeSeries,
  propTypes: propTypes,
  proxyChildren: proxyChildren,
  transform: transform,
  value: value,
  transforms: transforms
};
exports["default"] = _default;
