// rumble-charts/lib/helpers/normalizeSeries.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = normalizeSeries;

var _omitBy2 = _interopRequireDefault(require("lodash/omitBy"));

var _isEmpty2 = _interopRequireDefault(require("lodash/isEmpty"));

var _defaults2 = _interopRequireDefault(require("lodash/defaults"));

var _isArray2 = _interopRequireDefault(require("lodash/isArray"));

var _isNumber2 = _interopRequireDefault(require("lodash/isNumber"));

var _map2 = _interopRequireDefault(require("lodash/map"));

var _isUndefined2 = _interopRequireDefault(require("lodash/isUndefined"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var isInvalidLimit = function isInvalidLimit(value) {
  return (0, _isUndefined2["default"])(value) || value === Infinity || value === -Infinity;
};

function normalizeSeries(props) {
  var maxX = -Infinity,
      maxY = -Infinity,
      minX = Infinity,
      minY = Infinity;
  var series = (0, _map2["default"])(props.series, function (series) {
    var data = (0, _map2["default"])(series.data, function (item, index) {
      var d;

      if (!props.seriesNormalized) {
        d = {};

        if ((0, _isNumber2["default"])(item)) {
          d.x = index;
          d.y = item;
        } else if ((0, _isArray2["default"])(item)) {
          d.x = item[0];
          d.y = item[1];
        } else {
          d = item || {};

          if ((0, _isUndefined2["default"])(d.x)) {
            d.x = index;
          }
        }
      } else {
        d = item;
      }

      if ((0, _isUndefined2["default"])(props.maxX)) {
        maxX = Math.max(maxX, d.x || 0);
      }

      if ((0, _isUndefined2["default"])(props.maxY)) {
        maxY = Math.max(maxY, d.y || 0);
      }

      if ((0, _isUndefined2["default"])(props.minX)) {
        minX = Math.min(minX, d.x || 0);
      }

      if ((0, _isUndefined2["default"])(props.minY)) {
        minY = Math.min(minY, d.y || 0);
      }

      return d;
    });
    return (0, _defaults2["default"])({
      data: data
    }, series);
  });

  if ((0, _isEmpty2["default"])(series)) {
    series = undefined;
  }

  if (!(0, _isUndefined2["default"])(props.maxX)) {
    maxX = props.maxX;
  }

  if (!(0, _isUndefined2["default"])(props.maxY)) {
    maxY = props.maxY;
  }

  if (!(0, _isUndefined2["default"])(props.minX)) {
    minX = props.minX;
  }

  if (!(0, _isUndefined2["default"])(props.minY)) {
    minY = props.minY;
  }

  return (0, _omitBy2["default"])({
    seriesNormalized: true,
    series: series,
    maxX: maxX,
    maxY: maxY,
    minX: minX,
    minY: minY
  }, isInvalidLimit);
}
