// rumble-charts/lib/helpers/propTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.series = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var series = _propTypes["default"].arrayOf(_propTypes["default"].shape({
  name: _propTypes["default"].string,
  color: _propTypes["default"].string,
  opacity: _propTypes["default"].number,
  style: _propTypes["default"].object,
  data: _propTypes["default"].arrayOf(_propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].arrayOf(_propTypes["default"].number), _propTypes["default"].shape({
    x: _propTypes["default"].number,
    y: _propTypes["default"].number,
    color: _propTypes["default"].string,
    opacity: _propTypes["default"].number,
    style: _propTypes["default"].object
  })]))
}));

exports.series = series;
var _default = {
  series: series
};
exports["default"] = _default;
